osu.Game/Rulesets/Mods/ModFailCondition.cs

```diff
@@ -19,12 +19,31 @@ public abstract class ModFailCondition : Mod, IApplicableToHealthProcessor, IApp
         public virtual bool PerformFail() => true;
 
         public virtual bool RestartOnFail => Restart.Value;
+        private HealthProcessor healthProcessorInternal;
 
         public void ApplyToHealthProcessor(HealthProcessor healthProcessor)
         {
+            healthProcessorInternal = healthProcessor;
```

Comment:
Ideally `ModFailCondition` should not store references to `HealthProcessor`s. Doing so runs the risk of future bugs caused by mod instances retaining previous health processor instances locally (i.e. adds unnecessary state).

I think the preferred solution would be to have `ModFailCondition` signal the failure to `HealthProcessor` (via event, or bindable, or something of the sort) rather than operate on it directly as done here.

Reply:
I may have misunderstood what you were proposing here. Still, I think my solution is comparable assuming `ApplyToHealthProcessor` is called when new health processors are created, which is the case as far as I know.

Reply:
I may have misunderstood what you were proposing here. Still, I think my solution is comparable assuming `ApplyToHealthProcessor` is called when new health processors are created, which is the case as far as I know.